packages/react-instantsearch-dom/src/core/createInstantSearchServer.js

```diff
@@ -59,9 +59,36 @@ const getSearchParameters = (indexName, searchParameters) => {
   };
 };
 
+/**
+ * The engine can return params: "query=xxx&query=yyy" if e.g. a query rule modifies it.
+ * This however will cause us to miss the cache hydration, so we make sure to keep
+ * only the first query (always the one from the parameters).
+ */
+function removeDuplicateQuery(params) {
+  let previousIndex = undefined;
+  const queryParamRegex = /&?query=[^&]*/g;
+  return params.replace(queryParamRegex, function replacer(match, offset) {
+    if (previousIndex !== undefined && offset > previousIndex) {
```

Comment:
Can `offset > previousIndex` be false when `previousIndex !== undefined`?

Reply:
if there's three instances of `query` that would be possible, so I left that condition to be sure

Reply:
When you meet the first instance of `query`, `previousIndex` gets assigned (no longer undefined). After that, no matter how many more instances of `query` occurs, we can just return an empty string, IIUC.
Meaning, we can have just a boolean instead of `previousIndex` to store the offset.

Reply:
good call, fixed